### PR TITLE
refactor: (ecam) Convert FCTL page to React

### PR DESCRIPTION
**Pull request by RichardPilbery**
Fixes #[issue_no]

## Summary of Changes

Refactor FCTL page to use Reach and new SimVars. Changes should be invisible to user.

## Screenshots (if necessary)

ECAM-FCTL-engine2
ECAM-fctl-off
ECAM-FCTL-right

## Additional context

Discord username (if different from GitHub):

## Testing instructions

1. Start engines
2. Check all control service movements (ailerons, elevators, rudder and spoilers) are shown in the ECAM
3. Check spoilers show when armed speedbrake deploys

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/instruments/src/SD/Pages/Fctl/Fctl.tsx [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+            leftorright="left"
+            index={index}
+            x={leftX}
+            y={YCoord}
+            yw={YCoordW}
+            hydAvail={[hydraulicGAvailable, hydraulicBAvailable, hydraulicYAvailable]}
+            speedbrake={spoilerHandleState}
+            spoilerpos={leftSpoilerState}
+            ailpos={aileronLeftDeflectionState}
+            spoilersArmed={spoilersArmed}
```
Comment (beheh) [inline src/instruments/src/SD/Pages/Fctl/Fctl.tsx:161]: To suppress some potential React warnings, pass `key={index}` here (whenever using JSX in an array).

File: src/instruments/src/SD/Pages/Fctl/Fctl.tsx [hunk 2]
```diff
@@ -0,0 +1,767 @@
...
+            leftorright="right"
+            index={index}
+            x={rightX}
+            y={YCoord}
+            yw={YCoordW}
+            hydAvail={[hydraulicGAvailable, hydraulicBAvailable, hydraulicYAvailable]}
+            speedbrake={spoilerHandleState}
+            spoilerpos={rightSpoilerState}
+            ailpos={aileronRightDeflectionState}
+            spoilersArmed={spoilersArmed}
```
Comment (beheh) [inline src/instruments/src/SD/Pages/Fctl/Fctl.tsx:173]: Also add `key={index}` here.

File: src/instruments/src/SD/Pages/Fctl/Fctl.tsx [hunk 3]
```diff
@@ -0,0 +1,767 @@
...
+                <text
+                    id="pageTitle"
+                    className="PageTitle"
+                    x="45"
+                    y="18"
+                    textAnchor="middle"
+                    alignmentBaseline="central"
+                >
+                    F/CTL
+
```
Comment (beheh) [inline src/instruments/src/SD/Pages/Fctl/Fctl.tsx:189]: Is this empty line really required?

File: src/instruments/src/SD/Pages/Fctl/Fctl.tsx [hunk 4]
```diff
@@ -0,0 +1,767 @@
...
+                    <text
+                        id="elacText"
+                        className="Note"
+                        x="195"
+                        y="178"
+                        textAnchor="middle"
+                        alignmentBaseline="central"
+                    >
+                        ELAC
+
```
Comment (beheh) [inline src/instruments/src/SD/Pages/Fctl/Fctl.tsx:254]: Empty line.

File: src/instruments/src/SD/Pages/Fctl/Fctl.tsx [hunk 5]
```diff
@@ -0,0 +1,767 @@
...
+                    />
+                    <path
+                        id="rudderTail"
+                        className={hydraulicGAvailable || hydraulicBAvailable || hydraulicYAvailable ? 'GreenShape' : 'WarningShape'}
+                        d="M292,434 l8,48 l8,-48"
+                    />
+                </g>
+            </svg>
+        </>
+
```
Comment (beheh) [inline src/instruments/src/SD/Pages/Fctl/Fctl.tsx:455]:
```suggestion
```

**Review by beheh — CHANGES_REQUESTED** [review]

File: src/instruments/rollup.config.js [hunk 6]
```diff
@@ -41,8 +41,8 @@ const extraInstruments = [
         path: 'SD/Pages/Door',
     },
     {
-        name: 'cond-page',
-        path: 'SD/Pages/Cond',
+        name: 'fctl-page',
```
[the remark this answers is not in the document]
Comment (RichardPilbery) [reply src/instruments/rollup.config.js:44]: Good spot. I shouldn't have removed the ref to cond

File: src/instruments/src/SD/Pages/Fctl/Fctl.tsx (hunk 2 above)
Comment (RichardPilbery) [reply src/instruments/src/SD/Pages/Fctl/Fctl.tsx:173]: So the issue here is that this is not actually the index of the array, but a reference to the spoiler number. Poor choice of variable name.

**Review by juflaura — APPROVED** [review]

File: src/instruments/src/SD/Pages/Fctl/Fctl.tsx [hunk 7]
```diff
@@ -0,0 +1,761 @@
+/*
+ * A32NX
+ * Copyright (C) 2020 FlyByWire Simulations and its contributors
```
Comment (tracernz) [inline src/instruments/src/SD/Pages/Fctl/Fctl.tsx:3]: 2020?

**Review by beheh — APPROVED** [review]

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 31/3/21
Version of the sim: 1.14.6.0
PR Tested: #4193 
Tier of testing: 1
Changes to observe:
FCTL page in ECAM

Testing technique: Loaded in at gate, runway and approach
Steps you've made to test the Pull Request:
1. Loaded at gate 
(when the engines are off and you move the controls around they all move apart from the spoilers in the page and outside but the spoilers don't show as moving, I haven't included it as an issue as it works all the time just not in this instance and I think that's how it is normally as this is just converting it to React but if possible maybe that could be fixed unless that's how its supposed to be)
2. turned off flight control and tried
3. Loaded at runway
4. Loaded on approach

Media: Insert screenshots, recordings, or anything relevant to the review here.
image
image
image
image



Issues (If Any): None

Overall Rating: Good
Conclusions of the testing: LGTM

**Comment by DavidEGTE** [comment]
Category: QA Tester
Name: David#2177
Date of testing: 31/03/2021
Version of the sim: 1.14.6.0
PR Tested: #4193
Tier of testing: 1

Changes to observe:
(ecam) Convert FCTL page to React

Testing technique: Try-out at gate, runway and approach

Steps you've made to test the Pull Request:
1. Loaded at gate;
same issue as @FlightSimmer7700 https://github.com/flybywiresim/a32nx/pull/4193#issuecomment-811298259

2. Loaded at runway, works
3. Loaded on approach, works


Flight notes (When done a full flight):
N/A
Media: Insert screenshots, recordings, or anything relevant to the review here.
N/A
Issues (If Any): Write any issues
N/A
Steps to achieve the issue:
N/A

Overall Rating: Good
Conclusions of the testing: GTG

**Comment by pilotJesse** [comment]
Category: QA Trial
Name: Jesse#8659
Date of testing: 01/04/2021
Version the tester uses: 1.14.6.0
PR Tested: #5652
Tier of testing: 1
Changes to observe:
Refactor FCTL page to use Reach and new SimVars

Testing technique: 
1. Engines started on stand, tested full flight control check

Steps you've made to test the Pull Request:
1. Ailerons- Full Left, Full Right
2. Elevators- Full Down, Full Up
3. Rudders- Full Left, Full Right

Flight notes (When done a full flight):
Flight Control movements when engines started displaying correctly.

Media: 
EVL UP (2)
ELV DOWN
LEFT RUDDER
RIGHT RUDDER
RIGHT AIL
LEFT AIL


Issues (If Any): N/A

Overall Rating: Good
Conclusions of the testing: I have seen the above comments regarding a problem with the spoilers not displaying on the FTCL page. I only tested the above requirements thus on engine start.

**Review by lukecologne — CHANGES_REQUESTED** [review]
A few small issues (visually, not code-related):

- The ° sign on the pitch trim indicator should be cyan
- The gray box around the hydraulic system status indicators should not have rounded corners, see here: https://youtu.be/HBeFp08XG5M?t=253
- The rudder travel limit indicators paths seem to not be connected, resulting in the lines overlapping, that should not be the case (see the video above)
- A small missing feature, if you want it can be implemented later: The rudder indicator should have a cyan marker to indicate the rudder trim position

**Review by lukecologne — APPROVED** [review]

**Comment by lukecologne** [comment]
Alright, looks good now! Needs a rebase though after the fork

File: .husky/pre-commit [hunk 8]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
Comment (beheh) [inline .husky/pre-commit:1]: This was intentionally removed in master. You're reintroducing husky here.

**Review by beheh — CHANGES_REQUESTED** [review]

File: .husky/pre-commit (hunk 8 above)
Comment (RichardPilbery) [reply .husky/pre-commit:1]: Oh FFS. I only wanted to delete a chuffing CSS file.

**Comment by beheh** [comment]
In fact, due to the hard fork everything looks a bit weird right now. You might be better to cleanly pick up your changes again onto the new master.

**Comment by lukecologne** [comment]
It worked when I rebased it, after 6b1716b it's borked again

**Pull request closed without merging**